File: Linear_Control_Course/Chapter28/Lesson2/Chapter28_Lesson2.cpp
#include <iostream>
#include <vector>
#include <cmath>

struct State {
    double x1; // position y
    double x2; // velocity dy/dt
};

State deriv(const State& x, double t, double omega_n, double zeta, double u) {
    (void)t; // t not used in this time-invariant example
    State dx;
    dx.x1 = x.x2;
    dx.x2 = -2.0 * zeta * omega_n * x.x2
            - omega_n * omega_n * x.x1
            + omega_n * omega_n * u;
    return dx;
}

int main() {
    double omega_n = 20.0;
    double zeta    = 0.4;
    double h       = 0.001;   // step size [s]
    double t_final = 4.0;     // final time [s]

    int n_steps = static_cast<int>(t_final / h);
    std::vector<double> t(n_steps + 1);
    std::vector<double> y(n_steps + 1);

    State x{0.0, 0.0}; // initial conditions y(0) = 0, dy/dt(0) = 0

    for (int k = 0; k <= n_steps; ++k) {
        t[k] = k * h;
        y[k] = x.x1;

        double tk = t[k];
        double u  = 1.0; // unit-step input

        State k1 = deriv(x, tk,          omega_n, zeta, u);
        State k2 = deriv({x.x1 + 0.5*h*k1.x1, x.x2 + 0.5*h*k1.x2},
                         tk + 0.5*h, omega_n, zeta, u);
        State k3 = deriv({x.x1 + 0.5*h*k2.x1, x.x2 + 0.5*h*k2.x2},
                         tk + 0.5*h, omega_n, zeta, u);
        State k4 = deriv({x.x1 + h*k3.x1, x.x2 + h*k3.x2},
                         tk + h, omega_n, zeta, u);

        x.x1 += (h / 6.0) * (k1.x1 + 2.0*k2.x1 + 2.0*k3.x1 + k4.x1);
        x.x2 += (h / 6.0) * (k1.x2 + 2.0*k2.x2 + 2.0*k3.x2 + k4.x2);
    }

    // Export y[k] and t[k] to file or plot (e.g., using gnuplot or Python)
    std::cout << "# t  y\n";
    for (int k = 0; k <= n_steps; ++k) {
        std::cout << t[k] << " " << y[k] << "\n";
    }
    return 0;
}

File: Linear_Control_Course/Chapter28/Lesson2/Chapter28_Lesson2.java
public class SecondOrderSystem {

    private final double omegaN;
    private final double zeta;

    public SecondOrderSystem(double omegaN, double zeta) {
        this.omegaN = omegaN;
        this.zeta = zeta;
    }

    public static class Response {
        public final double[] t;
        public final double[] y;

        public Response(double[] t, double[] y) {
            this.t = t;
            this.y = y;
        }
    }

    // Simulate unit-step response y(t) for t in [0, tFinal] with step h
    public Response simulateStep(double tFinal, double h) {
        int nSteps = (int) Math.round(tFinal / h);
        double[] t = new double[nSteps + 1];
        double[] y = new double[nSteps + 1];

        double x1 = 0.0; // y
        double x2 = 0.0; // dy/dt

        for (int k = 0; k <= nSteps; ++k) {
            t[k] = k * h;
            y[k] = x1;

            double u = 1.0; // unit step
            double dx1 = x2;
            double dx2 = -2.0 * zeta * omegaN * x2
                         - omegaN * omegaN * x1
                         + omegaN * omegaN * u;

            x1 += h * dx1;
            x2 += h * dx2;
        }
        return new Response(t, y);
    }

    public static void main(String[] args) {
        SecondOrderSystem sys = new SecondOrderSystem(20.0, 0.4);
        Response resp = sys.simulateStep(4.0, 0.001);

        // Print first few samples (can be plotted in MATLAB, Python, etc.)
        for (int k = 0; k < resp.t.length; ++k) {
            System.out.println(resp.t[k] + " " + resp.y[k]);
        }
    }
}
